src/main/java/com/parasoft/findings/sonar/soatest/SurefireSOAtestParser.java

```diff
@@ -0,0 +1,150 @@
+
+    public SurefireSOAtestParser(FileSystem fs) {
+        this.fs = fs;
+    }
+
+    public void collect(SensorContext context, List<File> xunitFiles) {
+        parseFiles(context, xunitFiles);
+    }
+
+    private void parseFiles(SensorContext context, List<File> reports) {
+        UnitTestIndex index = new UnitTestIndex();
+        parseFiles(reports, index);
+        sanitize(index);
+        save(index, context);
+    }
+
+    private static void parseFiles(List<File> reports, UnitTestIndex index) {
+        StaxParser parser = new StaxParser(index);
+        for (File report : reports) {
+            try {
+                parser.parse(report);
+            } catch (XMLStreamException e) {
+                throw new InvalidReportException(NLS.getFormatted(Messages.FailToParseSurefireReport, report), e);
+            }
+        }
+    }
+
+    private static void sanitize(UnitTestIndex index) {
+        for (String classname : index.getClassnames()) {
+            if (StringUtils.contains(classname, "$")) {
+                // Surefire reports classes whereas sonar supports files
+                String parentClassName = StringUtils.substringBefore(classname, "$");
+                index.merge(classname, parentClassName);
+            }
+        }
+    }
+
+    private void save(UnitTestIndex index, SensorContext context) {
+        long negativeTimeTestNumber = 0;
+        Map<InputFile, UnitTestClassReport> indexByInputFile = mapToInputFile(index.getIndexByClassname());
+        for (Map.Entry<InputFile, UnitTestClassReport> entry : indexByInputFile.entrySet()) {
+            UnitTestClassReport report = entry.getValue();
+            if (report.getTests() > 0) {
+                negativeTimeTestNumber += report.getNegativeTimeTestNumber();
+                save(report, entry.getKey(), context);
+            }
+        }
+        if (negativeTimeTestNumber > 0) {
+            Logger.getLogger().warn(NLS.getFormatted(Messages.TotalDurationNotAccurateWithNegativeTimeTests, negativeTimeTestNumber));
+        }
+    }
+
+    private Map<InputFile, UnitTestClassReport> mapToInputFile(Map<String, UnitTestClassReport> indexByClassname) {
+        Map<InputFile, UnitTestClassReport> result = new HashMap<>();
+        indexByClassname.forEach((className, index) -> {
+            InputFile resource = getUnitTestResource(className, index);
+            if (resource != null) {
+                UnitTestClassReport report = result.computeIfAbsent(resource, r -> new UnitTestClassReport());
+                // in case of repeated/parameterized tests (JUnit 5.x) we may end up with tests having the same name
+                index.getResults().forEach(report::add);
+            } else {
+                Logger.getLogger().debug(NLS.getFormatted(Messages.ResourceNotFound, className));
+            }
+        });
+        return result;
+    }
+
+    private static void save(UnitTestClassReport report, InputFile inputFile, SensorContext context) {
```

Comment:
I see CoreMetrics.TEST_SUCCESS_DENSITY(success rate) is not set here. Will it display the correct success rate?